src/KOKKOS/fix_efield_kokkos.cpp

```diff
@@ -167,144 +165,186 @@ void FixEfieldKokkos<DeviceType>::post_force(int /*vflag*/)
-    //Kokkos::parallel_reduce(Kokkos::RangePolicy<DeviceType, TagFixEfieldNonConstant>(0,nlocal),*this,fsum_kk);
-    {
-    // local variables for lambda capture
-    auto prd = Few<double,3>(domain->prd);
-    auto h = Few<double,6>(domain->h);
-    auto triclinic = domain->triclinic;
-    auto l_ex = ex;
-    auto l_ey = ey;
-    auto l_ez = ez;
-    auto l_d_efield = d_efield;
-
-    auto l_x = x;
-    auto l_q = q;
-    auto l_f = f;
-    auto l_mask = mask;
-    auto l_image = image;
-    auto l_groupbit = groupbit;
-    auto l_xstyle = xstyle;
-    auto l_ystyle = ystyle;
-    auto l_zstyle = zstyle;
-
-    Kokkos::parallel_reduce(nlocal, LAMMPS_LAMBDA(const int& i, double_4& fsum_kk) {
-      if (l_mask[i] & l_groupbit) {
-        Few<double,3> x_i;
-        x_i[0] = l_x(i,0);
-        x_i[1] = l_x(i,1);
-        x_i[2] = l_x(i,2);
-        auto unwrap = DomainKokkos::unmap(prd,h,triclinic,x_i,l_image(i));
-        auto qtmp = l_q(i);
-        auto fx = qtmp * l_ex;
-        auto fy = qtmp * l_ey;
-        auto fz = qtmp * l_ez;
-        if (l_xstyle == ATOM) l_f(i,0) += qtmp * l_d_efield(i,0);
-        else if (l_xstyle) l_f(i,0) += fx;
-        if (l_ystyle == ATOM) l_f(i,1) += qtmp * l_d_efield(i,1);
-        else if (l_ystyle) l_f(i,1) += fy;
-        if (l_zstyle == ATOM) l_f(i,2) += qtmp * l_d_efield(i,2);
-        else if (l_zstyle) l_f(i,2) += fz;
-        fsum_kk.d0 -= fx * unwrap[0] + fy * unwrap[1] + fz * unwrap[2];
-        fsum_kk.d1 += fx;
-        fsum_kk.d2 += fy;
-        fsum_kk.d3 += fz;
-      }
-    },fsum_kk);
-    }
+
+    if(qflag && muflag)
+      Kokkos::parallel_reduce(Kokkos::RangePolicy<DeviceType, TagFixEfieldNonConstant<1,1> >(0,nlocal),*this,result);
+    else if(qflag && !muflag)
+      Kokkos::parallel_reduce(Kokkos::RangePolicy<DeviceType, TagFixEfieldNonConstant<1,0> >(0,nlocal),*this,result);
+    else if(!qflag && muflag)
+      Kokkos::parallel_reduce(Kokkos::RangePolicy<DeviceType, TagFixEfieldNonConstant<0,1> >(0,nlocal),*this,result);
+    else
+      Kokkos::parallel_reduce(Kokkos::RangePolicy<DeviceType, TagFixEfieldNonConstant<0,0> >(0,nlocal),*this,result);
 
     copymode = 0;
+
+  }
+
+
+  atomKK->modified(execution_space, datamask_modify);
+
+  Kokkos::atomic_store(&(d_fsum[0]),result[0]);
+  Kokkos::atomic_store(&(d_fsum[1]),result[1]);
+  Kokkos::atomic_store(&(d_fsum[2]),result[2]);
+  Kokkos::atomic_store(&(d_fsum[3]),result[3]);
+  k_fsum.template modify<DeviceType>();
+  k_fsum.template sync<LMPHostType>();
+
+  if (vflag_global) {
+    virial[0] += result[4];
+    virial[1] += result[5];
+    virial[2] += result[6];
+    virial[3] += result[7];
+    virial[4] += result[8];
+    virial[5] += result[9];
+  }
+
+  if (vflag_atom) {
+    k_vatom.template modify<DeviceType>();
+    k_vatom.template sync<LMPHostType>();
   }
 
-  atomKK->modified(execution_space, F_MASK);
 
-  fsum[0] = fsum_kk.d0;
```

Comment:
@alphataubio I don't understand why this doesn't work, I think it should. The value from a parallel_reduce is automatically transferred to host CPU, it should be current.

Reply:
it does now. this was my first attempt at debugging failed test but instead it was a symptom of another issue https://github.com/lammps/lammps/issues/4257.

along the way i ended up:
- fixing variable case (TagFixEfieldNonConstant operator) which didnt work anyways in kokkos class
- adding `fix_modify virial yes` support to match documentation, which was missing in kokkos class
- adding dipole support in kokkos class after i noticed your dipole code in base class, which was missing in kokkos class